xc/device.ml

```diff
@@ -2785,6 +2785,7 @@ module Dm_Common = struct
         (List.concat
            [
              [sprintf "file=%s" file; "if=none"; sprintf "id=%s" id]
+           ; (if String.length file > 0 then ["auto-read-only=off"] else [])
```

Comment:
You mean a CDROM disappears during migration?

Reply:
I think it's more idiomatic to test `file <> ""`. But `String.length` is O(1) (unlike in C), so it's not more expensive.

Reply:
If `file` can be the empty string: we have above `"file=%s"` - is this not a problem?

Reply:
> You mean a CDROM disappears during migration?
No mention of cdrom anywhere ;-). Actually cdrom are fine. The issue is with hard-drives which the guest can modify. The problem is that QEMU opens drive images as read-only, then chroot itself then find out it needs write permission to some images so attempt to reopen the image, but those aren't available in the chroot.
Since we know which images are to be open read-only, we can tell qemu to stop been smart and open the images with the permission needed, depending on the read-only property given (if not given, it default to read-only=off).

Reply:
> I think it's more idiomatic to test `file <> ""`. But `String.length` is O(1) (unlike in C), so it's not more expensive.

Is `file <> ""` true when `file` is an empty string?
If so, should I replace the condition by `!(file <> "")` or should I invert the code in the alternative like this `if file <> "" then [] else ["auto-read-only=off"]`?

Reply:
> If `file` can be the empty string: we have above `"file=%s"` - is this not a problem?

That doesn't seems to be an issue for QEMU. This also happen with libxl upstream when adding an empty cdrom drive.

Reply:
> Is file <> "" true when file is an empty string?

No.

What exactly is the relationship between the `file` argument and "hard-drives which the guest can modify"?

> (if not given, it default to read-only=off).

So `read-only=off` is the opposite of `auto-read-only=off`? I'm a bit confused :)

Reply:
> So `read-only=off` is the opposite of `auto-read-only=off`? I'm a bit confused :)

`auto-read-only` is a bit weird and confusing yes :-). But those two properties are different.
I'm quoting the doc (https://www.qemu.org/docs/master/system/invocation.html):
```
read-only
    Open the node read-only. Guest write attempts will fail.
auto-read-only
    If auto-read-only=on is set, QEMU may fall back to read-only usage even
    when read-only=off is requested, or even switch between modes as needed,
    e.g. depending on whether the image file is writable or whether a writing
    user is attached to the node.
```
Hope that helps a bit.

Reply:
String comparison in OCaml:
* `str = ""` is true if the string is empty
* `str <> ""` is true when the string is not empty
* `!` is not logical not, you would have to use `not` for that
* `String.length` returns the number of bytes in the string but it's not a good choice if you only want to test for a string being empty or not